File: examples1/files.hs
module Main (main) where

import Control.Exception (
    Exception (displayException),
    SomeException,
    catch,
 )
import System.Directory (getHomeDirectory)
import System.Exit (exitFailure)
import System.FilePath (joinPath)
import System.IO (hPutStrLn, stderr)

main :: IO ()
main = do
    homeDir <- getHomeDirectory
    let filePath = joinPath [homeDir, ".bashrc"]
    result <- readFile filePath `catch` handleErr
    putStr result `catch` handleErr

handleErr :: SomeException -> IO a
handleErr e = do
    hPutStrLn stderr (displayException e)
    exitFailure

File: examples1/files.rs
use std::env;
use std::fs;
use std::path;

pub fn main() {
    let home_dir = env::home_dir();
    let file_path = path::PathBuf::from([home_dir, ".bashrc"]);
    let file = handle_err(fs::read(file_path));
    let result = handle_err(std::str::from_utf8(file));
    println!("{result}")
}

fn handle_err<T, E>(v: Result<T, E>) -> T {
    match v {
        Ok(t) => t,
        Err(e) => {
            println!("{e}");
            std::process::exit(1)
        }
    }
}
